doc/notes/9947_type_decl_int.md

```diff
@@ -0,0 +1,235 @@
+which types reference which other types.
+
+5.3 Model of `model`
+
+There is a model in the `mcshared` project called `model`. This model currently
+consists of a single class with two options that get set by command line flags.
+The class itself is a singleton class and exists only to carry provide command
+line options to the model. Currently `m2x` and `x2m` are the only projects which
+use this model.
+
+5.3.1 Extending `model`
+
+For this work it was decided to provide a command line option to either emit all
+type forward declarations or emit no type forward declarations. Since `masl` was
+going to need to access a command line option, the `model` model would need to
+be extended. It was decided to extend this model to work for generic options, so
+that it would not have to continue to be updated for each new command line
+option.
+
+The new model is as follows:
+```
+|---------------------------------------------|                            |---------------------------------|
+| model                                       |                            | option                          |
+|---------------------------------------------|                            |---------------------------------|
+| name: string                            {I} | 1          R9200         * | model_name: string   {I, R9200} |
+|                                             |----------------------------| name: string                {I} |
+|---------------------------------------------| configures         applies | value: string                   |
+| void setoption( name:string, value:string ) |                            |                                 |
+| string getoption( name:string )             |                            |---------------------------------|
+| inst_ref<model> create( name:string )       |
+|---------------------------------------------|
+```
+
+`create` is a class based operation that creates a new instance of the model
+class. By convention, the "name" of the model should be the name of the project.
+
+5.4 Code changes
+
+5.4.1 Updated `model` model to work for generic options (see section 5.3.1)  
+5.4.2 Modified MASL to emit type forward declarations for every type if there
+exists an option with name "emittypeforwarddeclarations" and value "true".  
+5.4.3 Update references in `m2x` and `x2m` to old `model` API  
+5.4.4 Add command line option to `masl` to set the option mentioned in 5.4.2  
+5.4.5 Update the `xtumlmc_build` script to allow the new option  
+
+### 6. Implementation Comments
+
+6.1 Effect on MASL round trip
+
+Because `masl` will now default to always emit type forward declarations, MASL
+round trip tests will fail because the input models did not have the forward
+declarations in all cases. This will be solved simply by updating the test
+models to include type forward declarations everywhere.
+
+### 7. Unit Test
+
+7.1 MASL round trip shall pass
+
```

Comment:
Unit test should include eclipse-based test of the original error case

Reply:
done. test model attached to issue